File: 1. Creational Design Patterns/Abstract Factory/Java/AbstractFactoryExample.java
// Abstract Factory Pattern – Java implementation
// Compile: javac AbstractFactoryExample.java
// Run: java AbstractFactoryExample [mac]

interface Button {
    void paint();
}

interface Checkbox {
    void paint();
}

class WindowsButton implements Button {
    public void paint() {
        System.out.println("Render a button in Windows style.");
    }
}

class MacButton implements Button {
    public void paint() {
        System.out.println("Render a button in macOS style.");
    }
}

class WindowsCheckbox implements Checkbox {
    public void paint() {
        System.out.println("Render a checkbox in Windows style.");
    }
}

class MacCheckbox implements Checkbox {
    public void paint() {
        System.out.println("Render a checkbox in macOS style.");
    }
}

interface GUIFactory {
    Button createButton();
    Checkbox createCheckbox();
}

class WindowsFactory implements GUIFactory {
    public Button createButton() { return new WindowsButton(); }
    public Checkbox createCheckbox() { return new WindowsCheckbox(); }
}

class MacFactory implements GUIFactory {
    public Button createButton() { return new MacButton(); }
    public Checkbox createCheckbox() { return new MacCheckbox(); }
}

class Application {
    private final Button button;
    private final Checkbox checkbox;
    Application(GUIFactory factory) {
        button = factory.createButton();
        checkbox = factory.createCheckbox();
    }
    void paint() {
        button.paint();
        checkbox.paint();
    }
}

public class AbstractFactoryExample {
    public static void main(String[] args) {
        GUIFactory factory;
        if (args.length > 0 && args[0].equalsIgnoreCase("mac"))
            factory = new MacFactory();
        else
            factory = new WindowsFactory();
        Application app = new Application(factory);
        app.paint();
    }
}

File: 1. Creational Design Patterns/Abstract Factory/Kotlin/AbstractFactoryExample.kt
// Abstract Factory Pattern – Kotlin Example
// Compile & run: kotlinc AbstractFactoryExample.kt -include-runtime -d example.jar && java -jar example.jar [mac]

// Product A
interface Button {
    fun paint()
}

// Product B
interface Checkbox {
    fun paint()
}

// Concrete Products
class WindowsButton : Button {
    override fun paint() = println("Render a button in Windows style.")
}

class MacButton : Button {
    override fun paint() = println("Render a button in macOS style.")
}

class WindowsCheckbox : Checkbox {
    override fun paint() = println("Render a checkbox in Windows style.")
}

class MacCheckbox : Checkbox {
    override fun paint() = println("Render a checkbox in macOS style.")
}

// Abstract Factory
interface GUIFactory {
    fun createButton(): Button
    fun createCheckbox(): Checkbox
}

// Concrete Factories
class WindowsFactory : GUIFactory {
    override fun createButton() = WindowsButton()
    override fun createCheckbox() = WindowsCheckbox()
}

class MacFactory : GUIFactory {
    override fun createButton() = MacButton()
    override fun createCheckbox() = MacCheckbox()
}

// Client
class Application(factory: GUIFactory) {
    private val button: Button = factory.createButton()
    private val checkbox: Checkbox = factory.createCheckbox()
    fun paint() {
        button.paint()
        checkbox.paint()
    }
}

fun main(args: Array<String>) {
    val factory: GUIFactory = if (args.isNotEmpty() && args[0].equals("mac", ignoreCase = true)) MacFactory() else WindowsFactory()
    Application(factory).paint()
}
